src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -53,92 +56,43 @@ internal class ProjectBuildManager
-            MSB.Evaluation.ProjectCollection projectCollection,
-            CancellationToken cancellationToken)
+        public ProjectBuildManager(ImmutableDictionary<string, string> additionalGlobalProperties)
         {
-            var loadedProjects = projectCollection.GetLoadedProjects(path);
-            if (loadedProjects == null || loadedProjects.Count == 0)
-            {
-                return null;
-            }
-
-            // We need to walk through all of the projects that have been previously loaded from this path and
-            // find the one that has the given set of global properties, plus the default global properties that
-            // we load every project with.
-
-            globalProperties = globalProperties ?? ImmutableDictionary<string, string>.Empty;
-            var totalGlobalProperties = projectCollection.GlobalProperties.Count + globalProperties.Count;
-
-            foreach (var loadedProject in loadedProjects)
-            {
-                cancellationToken.ThrowIfCancellationRequested();
-
-                // If this project has a different number of global properties than we expect, it's not the
-                // one we're looking for.
-                if (loadedProject.GlobalProperties.Count != totalGlobalProperties)
-                {
-                    continue;
-                }
-
-                // Since we loaded all of them, the projects in this collection should all have the default
-                // global properties (i.e. the ones in _projectCollection.GlobalProperties). So, we just need to
-                // check the extra global properties.
-
-                var found = true;
-                foreach (var (key, value) in globalProperties)
-                {
-                    // MSBuild escapes the values of a project's global properties, so we must too.
-                    var escapedValue = MSB.Evaluation.ProjectCollection.Escape(value);
-
-                    if (!loadedProject.GlobalProperties.TryGetValue(key, out var actualValue) ||
-                        !string.Equals(actualValue, escapedValue, StringComparison.Ordinal))
-                    {
-                        found = false;
-                        break;
-                    }
-                }
-
-                if (found)
-                {
-                    return loadedProject;
-                }
-            }
-
-            // We couldn't find a project with this path and the set of global properties we expect.
-            return null;
+            _additionalGlobalProperties = additionalGlobalProperties ?? ImmutableDictionary<string, string>.Empty;
         }
 
-        public async Task<(MSB.Evaluation.Project project, DiagnosticLog log)> LoadProjectAsync(
-            string path, IDictionary<string, string> globalProperties, CancellationToken cancellationToken)
+        private ImmutableDictionary<string, string> AllGlobalProperties()
```

Comment:
Why a method?

Reply:
I think I had fantasies of a more significant computation originally. I agree that it's better as a property.